contracts/verifying/VerifyingSingletonPaymasterV2.sol

```diff
@@ -1,9 +1,363 @@
+        )
+            revert InsufficientBalance(
+                (requiredPreFund * dynamicMarkup) / PRICE_DENOMINATOR,
+                paymasterIdBalances[paymasterId]
+            );
+
+        context = abi.encode(
+            account,
+            paymasterId,
+            priceMarkup,
+            userOp.maxFeePerGas,
+            userOp.maxPriorityFeePerGas,
+            userOpHash
+        );
+
+        return (context, _packValidationData(false, validUntil, validAfter));
+    }
+
+    function parsePaymasterAndData(
+        bytes calldata paymasterAndData
+    )
+        public
+        pure
+        returns (
+            address paymasterId,
+            uint48 validUntil,
+            uint48 validAfter,
+            uint32 priceMarkup,
+            bytes calldata signature
+        )
+    {
+        (paymasterId, validUntil, validAfter, priceMarkup) = abi.decode(
+            paymasterAndData[VALID_PND_OFFSET:SIGNATURE_OFFSET],
+            (address, uint48, uint48, uint32)
+        );
+
+        signature = paymasterAndData[SIGNATURE_OFFSET:];
+    }
+
+    /**
+     * @dev Executes the paymaster's payment conditions
+     * @param mode tells whether the op succeeded, reverted, or if the op succeeded but cause the postOp to revert
+     * @param context payment conditions signed by the paymaster in `validatePaymasterUserOp`
+     * @param actualGasCost amount to be paid to the entry point in wei
+     */
+    function _postOp(
+        PostOpMode mode,
+        bytes calldata context,
+        uint256 actualGasCost
+    ) internal virtual override {
+        // Review what is needed in context
+        (
+            address account,
+            address paymasterId,
+            uint32 priceMarkup,
+            uint256 maxFeePerGas,
+            uint256 maxPriorityFeePerGas,
+            bytes32 userOpHash
+        ) = abi.decode(
+                context,
+                (address, address, uint32, uint256, uint256, bytes32)
+            );
+
+        uint256 effectiveGasPrice = getGasPrice(
+            maxFeePerGas,
+            maxPriorityFeePerGas
+        );
+
+        // Review max or min
+        uint32 dynamicMarkup = maxuint32(priceMarkup, fixedPriceMarkup);
+
+        uint256 balToDeduct = actualGasCost +
+            unaccountedEPGasOverhead *
+            effectiveGasPrice;
+
+        // deduct with premium
+        paymasterIdBalances[paymasterId] =
+            paymasterIdBalances[paymasterId] -
+            ((balToDeduct * dynamicMarkup) / PRICE_DENOMINATOR);
+
+        // "collect" premium
+        paymasterIdBalances[owner()] =
+            paymasterIdBalances[owner()] +
+            ((balToDeduct * (dynamicMarkup - 1e6)) / PRICE_DENOMINATOR);
+
+        emit GasBalanceDeducted(
+            paymasterId,
+            (balToDeduct * dynamicMarkup) / PRICE_DENOMINATOR
+        );
+        emit FeeCollected(
+            (balToDeduct * (dynamicMarkup - 1e6)) / PRICE_DENOMINATOR
+        );
```

Comment:
`balToDeduct * (dynamicMarkup - 1e6)) / PRICE_DENOMINATOR`

`(balToDeduct * dynamicMarkup) / PRICE_DENOMINATOR`

could be saved in memory vars

Reply:
they are only used once though

Reply:
I see. done